### PR TITLE
[SDF-20] Automatic Semantic Versioning

### PR DESCRIPTION
**Pull request by kubazuch**
Recommended reviewing order: `pull-request-labeler.yml` -> `pull-request-tagger.yml` -> `auto-release.yml`.

# Summary
This is a short summary of these three workflows.
## Stage 1: Labeling
Each opened PR will be analysed based on merging branch and a label will be applied accordingly:
- merge from `release/*` -> `version/major 🏗️`
- merge from `feature/*` -> `version/minor 🔨`
- merge from `bugfix/*` or `task/*` -> `version/patch 🪶`

> [!NOTE]
> I **do** realise, that this very PR breaks that rule, as I'm merging from `feature`, but I assigned a `patch` version, however I really didn't want to rename that branch xd
 
However, the programmer opening the PR can disobey this automatic assignment and change the labels however they want. The labeling workflow runs only once -- when PR is first opened.

## Stage 2: Tagging
When the PR is successfully merged, the second stage begins. Based on applied label a new tag is produced (with @kubazuch's access token) which is bumped based on the label. I won't explain which label does what because it is obvious.

> [!IMPORTANT]
> It is imperative that during this stage we do _**NOT**_ use github bot to create the tag. There is a peculiar security feature in GitHub, which prevents one bot-run-workflow to start another one. So if a bot created the tag, then the workflow reacting to creation of a tag would _**NOT**_ be triggered.

## Stage 3: Release
This stage is triggered by creation of a new tag. When a tag is created, we deploy a new release with proper changelog. However, there are no release artifacts, as I didn't want to duplicate code. This will be changed when other actions are refactored (i.e. not in the recent future).

### CODE REVIEW

**Comment by kubazuch** [comment]
The "Apply labels to Pull Requests 📝🏷️ / Label the PR with version label (pull_request)" workflow is failing because of GitHub security.

As we can read in `labeler`'s [README](https://github.com/actions/labeler?tab=readme-ov-file#permissions):
> In order to add labels to pull requests, the GitHub labeler action requires write permissions on the pull-request. However, when the action runs on a pull request from a forked repository, GitHub only grants read access tokens for pull_request events, at most. If you encounter an Error: HttpError: Resource not accessible by integration, it's likely due to these permission constraints.

We shall ignore this peculiar anomaly 😸🙂‍↕️

File: .github/workflows/pull-request-tagger.yml [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+    env:
+      MAJOR_LABEL: 'version/major 🏗️'
+      MINOR_LABEL: 'version/minor 🔨'
+      PATCH_LABEL: 'version/patch 🪶'
+    steps:
+      - name: Checkout code
+        id: checkout-code
+        uses: actions/checkout@v4 # I know this is a bad idea and I should do this prolly with GitHub api, this will be fixed later (i.e. never)
+        with:
+          # The binary versioning depends to on the git annotated tags.
```
Comment (migoox) [inline .github/workflows/pull-request-tagger.yml:24]:
```suggestion
          # The binary versioning depends on the git annotated tags. 
```

File: .github/workflows/auto-release.yml [hunk 2]
```diff
@@ -0,0 +1,32 @@
+name: Auto release 🍝
```
Comment (migoox) [inline .github/workflows/auto-release.yml:1]: I am not convinced that this step is required as in the current setup the auto release is invoked automatically when the tag is pushed, [see](https://github.com/gizmokis/resin/blob/7bd149616f5dd9b178c9beffca3bf54266b23573/.github/workflows/linux.yml#L68). If you wish you can replace the old strategy with the new one proposed here.

**Review by migoox — COMMENTED** [review]
Amazing job!

**Review by migoox — APPROVED** [review]

**Pull request merged**